src/OpenTelemetry.Instrumentation.ConfluentKafka/OpenTelemetryConsumeResultExtensions.cs

```diff
@@ -0,0 +1,161 @@
+    public static async ValueTask<ConsumeResult<TKey, TValue>?> ConsumeAndProcessMessageAsync<TKey, TValue>(
+        this IConsumer<TKey, TValue> consumer,
+        ConsumeAndProcessMessageHandler<TKey, TValue> handler,
+        CancellationToken cancellationToken = default)
+    {
+#if NETFRAMEWORK
+        if (consumer == null)
+        {
+            throw new ArgumentNullException(nameof(consumer));
+        }
+#else
+        ArgumentNullException.ThrowIfNull(consumer);
+#endif
+
+        if (consumer is not InstrumentedConsumer<TKey, TValue> instrumentedConsumer)
+        {
+            throw new ArgumentException("Invalid consumer type.", nameof(consumer));
+        }
+
+#if NETFRAMEWORK
+        if (handler is null)
+        {
+            throw new ArgumentNullException(nameof(handler));
+        }
+#else
+        ArgumentNullException.ThrowIfNull(handler);
+#endif
+
+        DateTimeOffset start = DateTimeOffset.UtcNow;
+
+        var consumeResult = instrumentedConsumer.Consume(cancellationToken);
+
+        if (consumeResult?.Message == null || consumeResult.IsPartitionEOF)
+        {
+            return consumeResult;
+        }
+
+        Activity? processActivity = null;
+        if (TryExtractPropagationContext(consumeResult, out var propagationContext))
+        {
+            TKey key = consumeResult.Message.Key;
+            object? keyAsObject = key;
+            processActivity = StartProcessActivity(
+                propagationContext,
+                start,
+                consumeResult.TopicPartitionOffset,
+                keyAsObject,
+                instrumentedConsumer.Name,
+                instrumentedConsumer.GroupId!);
+        }
+
+        try
+        {
+            await handler(consumeResult, cancellationToken).ConfigureAwait(false);
+        }
+        finally
+        {
+            processActivity?.Dispose();
+        }
+
+        return consumeResult;
+    }
+
+    internal static PropagationContext ExtractPropagationContext(Headers? headers)
+        => Propagators.DefaultTextMapPropagator.Extract(default, headers, ExtractTraceContext);
+
+    private static Activity? StartProcessActivity(PropagationContext propagationContext, DateTimeOffset start, TopicPartitionOffset? topicPartitionOffset, object? key, string clientId, string groupId)
+    {
+        var spanName = string.IsNullOrEmpty(topicPartitionOffset?.Topic)
+            ? ConfluentKafkaCommon.ProcessOperationName
+            : string.Concat(topicPartitionOffset!.Topic, " ", ConfluentKafkaCommon.ProcessOperationName);
+
+        ActivityLink[] activityLinks = propagationContext.ActivityContext.IsValid()
+            ? new[] { new ActivityLink(propagationContext.ActivityContext) }
+            : Array.Empty<ActivityLink>();
+
+        Activity? activity = ConfluentKafkaCommon.ActivitySource.StartActivity(spanName, kind: ActivityKind.Consumer, links: activityLinks, startTime: start, parentContext: default);
```

Comment:
you may (but don't have to) also set a parent to the same context as one if the message (if it's a single message). 
That's something most of the users seem to want, but some want the opposite.